### PR TITLE
`decode.rs`: Remove `unsafe` marker from safe functions

### PR DESCRIPTION
**Pull request by randomPoison**
Remove `unsafe` marker from the smaller functions in `decode.rs` that are already safe or mostly safe, and add `unsafe` blocks and safety comments where needed.

### CODE REVIEW

File: src/decode.rs [hunk 1]
```diff
@@ -3959,7 +3969,8 @@ unsafe fn setup_tile(
     ts.last_qidx = AtomicU8::new(frame_hdr.quant.yac);
     ts.last_delta_lf.fill_with(Default::default);
 
-    ts_c.msac = MsacContext::new(data, frame_hdr.disable_cdf_update != 0, &c.dsp.msac);
+    // SAFETY: `data` comes from `f.tiles`, which will live at least as long as the created `MsacContext`.
+    ts_c.msac = unsafe { MsacContext::new(data, frame_hdr.disable_cdf_update != 0, &c.dsp.msac) };
```
Comment (kkysen) [inline src/decode.rs:3973]: Not sure if we should make this one safe yet.  The lifetimes are not super clear and it might be possible to make this safe.

File: src/decode.rs [hunk 2]
```diff
@@ -790,14 +791,20 @@ unsafe fn read_pal_indices(
...
-    (pal_dsp.pal_idx_finish)(
-        pal_idx.unwrap_or(pal_tmp).as_mut_ptr(),
-        pal_tmp.as_ptr(),
-        bw4 as c_int * 4,
-        bh4 as c_int * 4,
-        w4 as c_int * 4,
-        h4 as c_int * 4,
-    );
+
+    // SAFETY: Unsafe asm call, checkasm tests have verified that the signature is
```
Comment (kkysen) [inline src/decode.rs:795]: Is that all `checkasm` does?

File: src/decode.rs [hunk 3]
```diff
@@ -790,14 +791,20 @@ unsafe fn read_pal_indices(
...
+    // correct. `pal_idx` and `pal_tmp` are at least `(bw4 * 4) * (bh4 * 4)`
+    // elements long, which is how many elements `pal_idx_finish` will read/write.
```
Comment (kkysen) [inline src/decode.rs:797]: These kinds of assertions should be `debug_assert!`s at the very least, I think.

File: src/refmvs.rs [hunk 4]
```diff
@@ -478,18 +478,28 @@ impl Rav1dRefmvsDSPContext {
...
         let bw4 = bw4 as _;
         let bh4 = bh4 as _;
-        (self.splat_mv)(rr.as_mut_ptr(), rmv, bx4, bw4, bh4, rr.len());
+
+        // SAFETY: Unsafe asm call, checkasm has verified the signature is correct. `rr` is at
+        // least `bh4` elements long, and each pointer in `rr` is non-null and points to at
+        // least `bx4 * bw4` elements, which is what will be accessed in `splat_mv`.
+        unsafe {
+            (self.splat_mv)(rr.as_mut_ptr(), rmv, bx4, bw4, bh4, rr.len());
+        }
```
Comment (kkysen) [inline src/refmvs.rs:502]: We pass `rr.len()` here, so the fallback `fn` uses that for the length.

**Review by kkysen — CHANGES_REQUESTED** [review]

File: src/decode.rs (hunk 2 above)
Comment (randomPoison) [reply src/decode.rs:795]: Naw, looks like the checkasm tests are full unit tests for the asm and C fallback routines, but since we're not running the checkasm tests against Rust the only thing it's verified for us is that the function signatures are correct. I'm open to better ways of wording this if you have suggestions.

File: src/decode.rs (hunk 1 above)
Comment (randomPoison) [reply src/decode.rs:3973]: I think marking `setup_tile` safe for now is fine, though if we can make `MsacContext::new` safe at a later point that'd be great.

File: src/decode.rs (hunk 2 above)
Comment (kkysen) [reply src/decode.rs:795]: If we're not running those tests, signature checking isn't even happening either right?  Can we just say eventually will be checked by checkasm?

File: src/decode.rs (hunk 1 above)
Comment (kkysen) [reply src/decode.rs:3973]: Hmm, it doesn't matter too much I guess.  It's fine as is then.

File: src/decode.rs (hunk 2 above)
Comment (randomPoison) [reply src/decode.rs:795]: Eh, I don't think it makes sense to say that will eventually be checked, but I can reword it to remove the reference to checkasm.

File: src/decode.rs (hunk 3 above)
Comment (randomPoison) [reply src/decode.rs:797]: Added. I also tweaked the wording of the safety comment since the required length of the two arrays is slightly different.

File: src/refmvs.rs (hunk 4 above)
Comment (randomPoison) [reply src/refmvs.rs]: Added that to the safety comment.

File: src/decode.rs [hunk 5]
```diff
@@ -797,14 +798,28 @@ unsafe fn read_pal_indices(
...
-    );
+
+    if let Some(pal_idx) = &pal_idx {
+        let read_len = bw4 * 2 * bh4 * 4;
+        debug_assert!(pal_idx.len() >= read_len);
+    }
+    let read_len = bw4 * 4 * bh4 * 4;
+    debug_assert!(pal_tmp.len() >= read_len);
+
+    // SAFETY: Unsafe asm call. `pal_idx` is at least `(bw * 2) * (bh * 4)`
```
Comment (kkysen) [inline src/decode.rs:809]:
```suggestion
    // SAFETY: Unsafe asm call. `pal_idx` is at least `(bw4 * 2) * (bh4 * 4)`
```

That's what `read_len` above is; one of these is wrong, right?

**Review by kkysen — APPROVED** [review]
I think one of the lengths in the comment has a typo?  Otherwise LGTM now.

File: src/decode.rs (hunk 5 above)
Comment (randomPoison) [reply src/decode.rs:809]: Ah yeah that should be `bw4` and `bh4`, will fix.
